### ISSUE TITLE
Pilot constantly sending full pushes

### ISSUE THREAD

**Issue by cainelli** [istio/istio#14605]
**Bug description**

From time to time random Pilot instances sends full pushes on every cycle causing high CPU usage and requires a restart to recover.

It seems to be a race condition or I'm falling into a special case with my k8s deployment. As I'm not familiar with Pilot code I've added additional logging [here](https://github.com/istio/istio/blob/1.1.7/pilot/pkg/proxy/envoy/v2/eds.go#L596-L608) which might be helpful :

```go
// 2. Update data for the specific cluster. Each cluster gets independent
// updates containing the full list of endpoints for the service in that cluster.
for _, e := range istioEndpoints {
  if e.ServiceAccount != "" {
    _, f = ep.ServiceAccounts[e.ServiceAccount]
    if !f && !internal {
      // The entry has a service account that was not previously associated.
      // Requires a CDS push and full sync.
      adsLog.Infof("Endpoint updating service account %s %s", e.ServiceAccount, serviceName)
      adsLog.Infof("endpoint: %v", e)
      adsLog.Infof("epServiceAccounts: %v", ep.ServiceAccounts)
      requireFull = true
    }
  }
}
```

```
2019-06-05T18:01:46.462758Z	info	ads	epServiceAccounts: map[spiffe://cluster.local/ns/kube-system/sa/calico-node:true]
2019-06-05T18:01:46.562945Z	info	ads	Push debounce stable[30] 1: 100.141193ms since last change, 100.141193ms since last push, full=true
2019-06-05T18:01:46.573340Z	info	ads	XDS: Pushing 2019-06-05T18:01:46Z/27 Services: 230, ConnectedEndpoints: 17
--
2019-06-05T18:01:48.072537Z	info	ads	Endpoint updating service account spiffe://cluster.local/ns/kube-system/sa/calico-node kubernetes.default.svc.cluster.local
2019-06-05T18:01:48.072566Z	info	ads	endpoint: &{map[k8s-app:calico-node pod-template-generation:1 controller-revision-hash:3080069882] tcp 10.20.26.106 443 https kubernetes://calico-node-px79w.kube-system <nil> spiffe://cluster.local/ns/kube-system/sa/calico-node   0}
2019-06-05T18:01:48.072575Z	info	ads	epServiceAccounts: map[spiffe://cluster.local/ns/kube-system/sa/:true]
2019-06-05T18:01:48.172769Z	info	ads	Push debounce stable[31] 1: 100.132521ms since last change, 100.132521ms since last push, full=true
2019-06-05T18:01:48.182334Z	info	ads	XDS: Pushing 2019-06-05T18:01:48Z/28 Services: 230, ConnectedEndpoints: 17
--
2019-06-05T18:01:56.477277Z	info	ads	Endpoint updating service account spiffe://cluster.local/ns/kube-system/sa/ kubernetes.default.svc.cluster.local
2019-06-05T18:01:56.477305Z	info	ads	endpoint: &{map[k8s-app:kube-apiserver tier:control-plane] tcp 10.20.9.86 443 https kubernetes://kube-apiserver-ip-10-20-9-86.eu-central-1.compute.internal.kube-system <nil> spiffe://cluster.local/ns/kube-system/sa/   0}
2019-06-05T18:01:56.477318Z	info	ads	epServiceAccounts: map[spiffe://cluster.local/ns/kube-system/sa/calico-node:true]
2019-06-05T18:01:56.577516Z	info	ads	Push debounce stable[32] 1: 100.148622ms since last change, 100.148622ms since last push, full=true
2019-06-05T18:01:56.591689Z	info	ads	XDS: Pushing 2019-06-05T18:01:56Z/29 Services: 230, ConnectedEndpoints: 17
--
2019-06-05T18:01:58.091679Z	info	ads	Endpoint updating service account spiffe://cluster.local/ns/kube-system/sa/calico-node kubernetes.default.svc.cluster.local
2019-06-05T18:01:58.091701Z	info	ads	endpoint: &{map[pod-template-generation:1 controller-revision-hash:3080069882 k8s-app:calico-node] tcp 10.20.26.106 443 https kubernetes://calico-node-px79w.kube-system <nil> spiffe://cluster.local/ns/kube-system/sa/calico-node   0}
2019-06-05T18:01:58.091712Z	info	ads	epServiceAccounts: map[spiffe://cluster.local/ns/kube-system/sa/:true]
2019-06-05T18:01:58.191936Z	info	ads	Push debounce stable[33] 1: 100.141933ms since last change, 100.141933ms since last push, full=true
2019-06-05T18:01:58.206305Z	info	ads	XDS: Pushing 2019-06-05T18:01:58Z/30 Services: 230, ConnectedEndpoints: 17
--
2019-06-05T18:02:06.493479Z	info	ads	Endpoint updating service account spiffe://cluster.local/ns/kube-system/sa/ kubernetes.default.svc.cluster.local
2019-06-05T18:02:06.493508Z	info	ads	endpoint: &{map[k8s-app:kube-apiserver tier:control-plane] tcp 10.20.9.86 443 https kubernetes://kube-apiserver-ip-10-20-9-86.eu-central-1.compute.internal.kube-system <nil> spiffe://cluster.local/ns/kube-system/sa/   0}
2019-06-05T18:02:06.493519Z	info	ads	epServiceAccounts: map[spiffe://cluster.local/ns/kube-system/sa/calico-node:true]
2019-06-05T18:02:06.593778Z	info	ads	Push debounce stable[34] 1: 100.180952ms since last change, 100.180952ms since last push, full=true
2019-06-05T18:02:06.603798Z	info	ads	XDS: Pushing 2019-06-05T18:02:06Z/31 Services: 230, ConnectedEndpoints: 17
--
2019-06-05T18:02:08.111740Z	info	ads	Endpoint updating service account spiffe://cluster.local/ns/kube-system/sa/calico-node kubernetes.default.svc.cluster.local
2019-06-05T18:02:08.111766Z	info	ads	endpoint: &{map[controller-revision-hash:3080069882 k8s-app:calico-node pod-template-generation:1] tcp 10.20.26.106 443 https kubernetes://calico-node-px79w.kube-system <nil> spiffe://cluster.local/ns/kube-system/sa/calico-node   0}
2019-06-05T18:02:08.111787Z	info	ads	epServiceAccounts: map[spiffe://cluster.local/ns/kube-system/sa/:true]
2019-06-05T18:02:08.212078Z	info	ads	Push debounce stable[35] 1: 100.218325ms since last change, 100.218325ms since last push, full=true
2019-06-05T18:02:08.222283Z	info	ads	XDS: Pushing 2019-06-05T18:02:08Z/32 Services: 230, ConnectedEndpoints: 17
--
2019-06-05T18:02:16.508897Z	info	ads	Endpoint updating service account spiffe://cluster.local/ns/kube-system/sa/ kubernetes.default.svc.cluster.local
2019-06-05T18:02:16.508925Z	info	ads	endpoint: &{map[tier:control-plane k8s-app:kube-apiserver] tcp 10.20.9.86 443 https kubernetes://kube-apiserver-ip-10-20-9-86.eu-central-1.compute.internal.kube-system <nil> spiffe://cluster.local/ns/kube-system/sa/   0}
2019-06-05T18:02:16.508937Z	info	ads	epServiceAccounts: map[spiffe://cluster.local/ns/kube-system/sa/calico-node:true]
2019-06-05T18:02:16.609173Z	info	ads	Push debounce stable[36] 1: 100.158153ms since last change, 100.158153ms since last push, full=true
2019-06-05T18:02:16.623490Z	info	ads	XDS: Pushing 2019-06-05T18:02:16Z/33 Services: 230, ConnectedEndpoints: 17
--
2019-06-05T18:02:18.128538Z	info	ads	Endpoint updating service account spiffe://cluster.local/ns/kube-system/sa/calico-node kubernetes.default.svc.cluster.local
2019-06-05T18:02:18.128557Z	info	ads	endpoint: &{map[controller-revision-hash:3080069882 k8s-app:calico-node pod-template-generation:1] tcp 10.20.26.106 443 https kubernetes://calico-node-px79w.kube-system <nil> spiffe://cluster.local/ns/kube-system/sa/calico-node   0}
2019-06-05T18:02:18.128567Z	info	ads	epServiceAccounts: map[spiffe://cluster.local/ns/kube-system/sa/:true]
2019-06-05T18:02:18.228815Z	info	ads	Push debounce stable[37] 1: 100.156687ms since last change, 100.156687ms since last push, full=true
2019-06-05T18:02:18.239285Z	info	ads	XDS: Pushing 2019-06-05T18:02:18Z/34 Services: 230, ConnectedEndpoints: 17
--
2019-06-05T18:02:26.530942Z	info	ads	Endpoint updating service account spiffe://cluster.local/ns/kube-system/sa/ kubernetes.default.svc.cluster.local
2019-06-05T18:02:26.530966Z	info	ads	endpoint: &{map[k8s-app:kube-apiserver tier:control-plane] tcp 10.20.9.86 443 https kubernetes://kube-apiserver-ip-10-20-9-86.eu-central-1.compute.internal.kube-system <nil> spiffe://cluster.local/ns/kube-system/sa/   0}
2019-06-05T18:02:26.530976Z	info	ads	epServiceAccounts: map[spiffe://cluster.local/ns/kube-system/sa/calico-node:true]
2019-06-05T18:02:26.631220Z	info	ads	Push debounce stable[38] 1: 100.164304ms since last change, 100.164304ms since last push, full=true
2019-06-05T18:02:26.641470Z	info	ads	XDS: Pushing 2019-06-05T18:02:26Z/35 Services: 230, ConnectedEndpoints: 17
--
2019-06-05T18:02:28.149966Z	info	ads	Endpoint updating service account spiffe://cluster.local/ns/kube-system/sa/calico-node kubernetes.default.svc.cluster.local
2019-06-05T18:02:28.149996Z	info	ads	endpoint: &{map[controller-revision-hash:3080069882 k8s-app:calico-node pod-template-generation:1] tcp 10.20.26.106 443 https kubernetes://calico-node-px79w.kube-system <nil> spiffe://cluster.local/ns/kube-system/sa/calico-node   0}
2019-06-05T18:02:28.150006Z	info	ads	epServiceAccounts: map[spiffe://cluster.local/ns/kube-system/sa/:true]
2019-06-05T18:02:28.250229Z	info	ads	Push debounce stable[39] 1: 100.154832ms since last change, 100.154832ms since last push, full=true
2019-06-05T18:02:28.261107Z	info	ads	XDS: Pushing 2019-06-05T18:02:28Z/36 Services: 230, ConnectedEndpoints: 17
--
2019-06-05T18:02:36.551617Z	info	ads	Endpoint updating service account spiffe://cluster.local/ns/kube-system/sa/ kubernetes.default.svc.cluster.local
2019-06-05T18:02:36.551643Z	info	ads	endpoint: &{map[k8s-app:kube-apiserver tier:control-plane] tcp 10.20.9.86 443 https kubernetes://kube-apiserver-ip-10-20-9-86.eu-central-1.compute.internal.kube-system <nil> spiffe://cluster.local/ns/kube-system/sa/   0}
2019-06-05T18:02:36.551654Z	info	ads	epServiceAccounts: map[spiffe://cluster.local/ns/kube-system/sa/calico-node:true]
2019-06-05T18:02:36.651976Z	info	ads	Push debounce stable[40] 1: 100.14281ms since last change, 100.14281ms since last push, full=true
2019-06-05T18:02:36.661784Z	info	ads	XDS: Pushing 2019-06-05T18:02:36Z/37 Services: 230, ConnectedEndpoints: 17
--
2019-06-05T18:02:38.169245Z	info	ads	Endpoint updating service account spiffe://cluster.local/ns/kube-system/sa/calico-node kubernetes.default.svc.cluster.local
2019-06-05T18:02:38.169268Z	info	ads	endpoint: &{map[controller-revision-hash:3080069882 k8s-app:calico-node pod-template-generation:1] tcp 10.20.26.106 443 https kubernetes://calico-node-px79w.kube-system <nil> spiffe://cluster.local/ns/kube-system/sa/calico-node   0}
2019-06-05T18:02:38.169279Z	info	ads	epServiceAccounts: map[spiffe://cluster.local/ns/kube-system/sa/:true]
2019-06-05T18:02:38.269476Z	info	ads	Push debounce stable[41] 1: 100.141845ms since last change, 100.141845ms since last push, full=true
2019-06-05T18:02:38.279755Z	info	ads	XDS: Pushing 2019-06-05T18:02:38Z/38 Services: 230, ConnectedEndpoints: 17
--
2019-06-05T18:02:46.567463Z	info	ads	Endpoint updating service account spiffe://cluster.local/ns/kube-system/sa/ kubernetes.default.svc.cluster.local
2019-06-05T18:02:46.567493Z	info	ads	endpoint: &{map[k8s-app:kube-apiserver tier:control-plane] tcp 10.20.9.86 443 https kubernetes://kube-apiserver-ip-10-20-9-86.eu-central-1.compute.internal.kube-system <nil> spiffe://cluster.local/ns/kube-system/sa/   0}
2019-06-05T18:02:46.567503Z	info	ads	epServiceAccounts: map[spiffe://cluster.local/ns/kube-system/sa/calico-node:true]
```

As you see above the data for serviceAccount of service `kubernetes.default.svc.cluster.local`  constantly changes between `map[spiffe://cluster.local/ns/kube-system/sa/:true]` and `map[spiffe://cluster.local/ns/kube-system/sa/calico-node:true]`  which hints some race condition by missing lock or something else.

Some observations:

* This happens in production and I can also reproduce in a sandbox cluster but I still don't know exactly in which conditions.
* We use static manifests for Kubernetes control plane and the API server Pods does not have a k8s service account.
* CPU load is not related to number of proxy instances a Pilot serves.
* When Pilot gets into this state is always involving `endpoint: &{map[k8s-app:kube-apiserver tier:control-plane]...` and `endpoint: &{map[controller-revision-hash:3080069882 k8s-app:calico-node pod-template-generation:1]...`. This happens in different clusters.
* I've checked similar issues like https://github.com/istio/istio/issues/14042 and can tell that there's no Pods crashing or endpoints moving from 0 to 1. But when I watch Endpoint resources in all namespaces I frequently get updates in some core components such as:
```
$ k get ep --all-namespaces -w
...
kube-system   kube-controller-manager   <none>    6m
kube-system   kube-scheduler   <none>    6m
kube-system   cluster-autoscaler   <none>    6m
default   kubernetes   10.20.26.106:443   7m
kube-system   kube-controller-manager   <none>    6m
kube-system   kube-scheduler   <none>    6m
kube-system   cluster-autoscaler   <none>    7m
kube-system   kube-controller-manager   <none>    6m
kube-system   kube-scheduler   <none>    6m
kube-system   cluster-autoscaler   <none>    7m
kube-system   kube-controller-manager   <none>    6m
kube-system   kube-scheduler   <none>    6m
kube-system   cluster-autoscaler   <none>    7m
kube-system   kube-controller-manager   <none>    6m
default   kubernetes   10.20.9.86:443   7m
kube-system   kube-scheduler   <none>    6m
kube-system   cluster-autoscaler   <none>    7m
kube-system   kube-controller-manager   <none>    7m
kube-system   kube-scheduler   <none>    6m
kube-system   cluster-autoscaler   <none>    7m
default   kubernetes   10.20.26.106:443   7m
kube-system   kube-controller-manager   <none>    7m
kube-system   kube-scheduler   <none>    7m
kube-system   cluster-autoscaler   <none>    7m
kube-system   kube-controller-manager   <none>    7m
kube-system   kube-scheduler   <none>    7m
kube-system   cluster-autoscaler   <none>    7m
kube-system   kube-controller-manager   <none>    7m
kube-system   kube-scheduler   <none>    7m
kube-system   cluster-autoscaler   <none>    7m
default   kubernetes   10.20.9.86:443   7m
kube-system   kube-controller-manager   <none>    7m
kube-system   kube-scheduler   <none>    7m
kube-system   cluster-autoscaler   <none>    7m
kube-system   kube-controller-manager   <none>    7m
kube-system   kube-scheduler   <none>    7m
kube-system   cluster-autoscaler   <none>    7m
default   kubernetes   10.20.26.106:443   7m
kube-system   kube-controller-manager   <none>    7m
kube-system   kube-scheduler   <none>    7m
kube-system   cluster-autoscaler   <none>    7m
kube-system   kube-controller-manager   <none>    7m
kube-system   kube-scheduler   <none>    7m
kube-system   cluster-autoscaler   <none>    7m
kube-system   kube-controller-manager   <none>    7m
kube-system   kube-scheduler   <none>    7m
kube-system   cluster-autoscaler   <none>    7m
default   kubernetes   10.20.9.86:443   7m
kube-system   kube-controller-manager   <none>    7m
kube-system   kube-scheduler   <none>    7m
```

Couriousily I don't have `Services` for kube-controller-manager, kube-scheduler or cluster-autoscaler and I can't tell if that's the normal behavior.

**Affected product area)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[x] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience

**Expected behavior**
Full pushes only when it applies.

**Steps to reproduce the bug**

**Version**
`istioctl version --remote`

```
client version: version.BuildInfo{Version:"1.1.6", GitRevision:"04850e14d38a69a38c16c800e237b1108056513e", User:"root", Host:"1844d064-72cc-11e9-a0d5-0a580a2c0304", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.5-11-g04850e1"}
ingressgateway version: version.BuildInfo{Version:"1.1.7", GitRevision:"eec7a74473deee98cad0a996f41a32a47dd453c2", User:"root", Host:"341b3bf0-76ac-11e9-b644-0a580a2c0404", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.6-6-geec7a74"}
ingressgateway version: version.BuildInfo{Version:"1.1.7", GitRevision:"eec7a74473deee98cad0a996f41a32a47dd453c2", User:"root", Host:"341b3bf0-76ac-11e9-b644-0a580a2c0404", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.6-6-geec7a74"}
ingressgateway version: version.BuildInfo{Version:"1.1.7", GitRevision:"eec7a74473deee98cad0a996f41a32a47dd453c2", User:"root", Host:"341b3bf0-76ac-11e9-b644-0a580a2c0404", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.6-6-geec7a74"}
citadel version: version.BuildInfo{Version:"1.1.7", GitRevision:"eec7a74473deee98cad0a996f41a32a47dd453c2-dirty", User:"root", Host:"341b3bf0-76ac-11e9-b644-0a580a2c0404", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.6-6-geec7a74"}
galley version: version.BuildInfo{Version:"1.1.7", GitRevision:"eec7a74473deee98cad0a996f41a32a47dd453c2-dirty", User:"root", Host:"341b3bf0-76ac-11e9-b644-0a580a2c0404", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.6-6-geec7a74"}
ingressgateway version: version.BuildInfo{Version:"1.1.7", GitRevision:"eec7a74473deee98cad0a996f41a32a47dd453c2", User:"root", Host:"341b3bf0-76ac-11e9-b644-0a580a2c0404", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.6-6-geec7a74"}
ingressgateway version: version.BuildInfo{Version:"1.1.7", GitRevision:"eec7a74473deee98cad0a996f41a32a47dd453c2", User:"root", Host:"341b3bf0-76ac-11e9-b644-0a580a2c0404", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.6-6-geec7a74"}
ingressgateway version: version.BuildInfo{Version:"1.1.7", GitRevision:"eec7a74473deee98cad0a996f41a32a47dd453c2", User:"root", Host:"341b3bf0-76ac-11e9-b644-0a580a2c0404", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.6-6-geec7a74"}
pilot version: version.BuildInfo{Version:"eec7a74473deee98cad0a996f41a32a47dd453c2-dirty", GitRevision:"eec7a74473deee98cad0a996f41a32a47dd453c2-dirty", User:"root", Host:"9908dd1e160b", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.7"}
pilot version: version.BuildInfo{Version:"eec7a74473deee98cad0a996f41a32a47dd453c2-dirty", GitRevision:"eec7a74473deee98cad0a996f41a32a47dd453c2-dirty", User:"root", Host:"9908dd1e160b", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.7"}
pilot version: version.BuildInfo{Version:"eec7a74473deee98cad0a996f41a32a47dd453c2-dirty", GitRevision:"eec7a74473deee98cad0a996f41a32a47dd453c2-dirty", User:"root", Host:"9908dd1e160b", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.7"}
policy version: version.BuildInfo{Version:"1.1.7", GitRevision:"eec7a74473deee98cad0a996f41a32a47dd453c2-dirty", User:"root", Host:"341b3bf0-76ac-11e9-b644-0a580a2c0404", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.6-6-geec7a74"}
policy version: version.BuildInfo{Version:"1.1.7", GitRevision:"eec7a74473deee98cad0a996f41a32a47dd453c2-dirty", User:"root", Host:"341b3bf0-76ac-11e9-b644-0a580a2c0404", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.6-6-geec7a74"}
policy version: version.BuildInfo{Version:"1.1.7", GitRevision:"eec7a74473deee98cad0a996f41a32a47dd453c2-dirty", User:"root", Host:"341b3bf0-76ac-11e9-b644-0a580a2c0404", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.6-6-geec7a74"}
sidecar-injector version: version.BuildInfo{Version:"1.1.7", GitRevision:"eec7a74473deee98cad0a996f41a32a47dd453c2-dirty", User:"root", Host:"341b3bf0-76ac-11e9-b644-0a580a2c0404", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.6-6-geec7a74"}
telemetry version: version.BuildInfo{Version:"1.1.7", GitRevision:"eec7a74473deee98cad0a996f41a32a47dd453c2-dirty", User:"root", Host:"341b3bf0-76ac-11e9-b644-0a580a2c0404", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.6-6-geec7a74"}
telemetry version: version.BuildInfo{Version:"1.1.7", GitRevision:"eec7a74473deee98cad0a996f41a32a47dd453c2-dirty", User:"root", Host:"341b3bf0-76ac-11e9-b644-0a580a2c0404", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.6-6-geec7a74"}
telemetry version: version.BuildInfo{Version:"1.1.7", GitRevision:"eec7a74473deee98cad0a996f41a32a47dd453c2-dirty", User:"root", Host:"341b3bf0-76ac-11e9-b644-0a580a2c0404", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.6-6-geec7a74"}
```

`kubectl version`
```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.3", GitCommit:"2bba0127d85d5a46ab4b778548be28623b32d0b0", GitTreeState:"clean", BuildDate:"2018-05-21T09:17:39Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.5", GitCommit:"32ac1c9073b132b8ba18aa830f46b77dcceb0723", GitTreeState:"clean", BuildDate:"2018-06-21T11:34:22Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```

**How was Istio installed?**
[Helm chart](https://gist.github.com/cainelli/fb4dff0bff9eaea9877874568933bc63). 

**Environment where bug was observed (cloud vendor, OS, etc)**
AWS, CoreOS

**Cluster Dump**

**Comment by cainelli** [comment]
Another observation: I noticed that `kubernetes.default.svc.cluster.local` was updating its endpoints every a couple of sec which increased the damage since on each update the full push was being sent. The reason is how the api server was configured, in 1.10 the default endpoint reconcile type is master count which I didn't specified and it was assuming 1. This caused a race condition with each master thinking it was the only one in the cluster. I've changed that and by adding `--endpoint-reconciler-type=lease` to the kubernetes api made things more stable and we see a drastic improve in CPU utilization because it's not doing full push every a couple of sec anymore.

![image](https://user-images.githubusercontent.com/3514984/59035145-ee74fa80-886c-11e9-8b8a-eb56bca6c259.png)

**Comment by howardjohn** [comment]
@cainelli thanks for the detailed investigation. 

I see a bunch of endpoint churn for kube-controller-manager, cluster-autoscaler, gcp-controller-manager, and kube-scheduler (on GKE) but not kubernetes.

Are you still seeing problems after fixing the kubernetes service, or do you still have problems with 

> As you see above the data for serviceAccount of service kubernetes.default.svc.cluster.local constantly changes between map[spiffe://cluster.local/ns/kube-system/sa/:true] and map[spiffe://cluster.local/ns/kube-system/sa/calico-node:true] which hints some race condition by missing lock or something else.

**Comment by cainelli** [comment]
Thank you for checking those endpoint updates in GKE as well, my guess is that they are used for leader election.

I'm afraid that the problem is still there and I'm not seeing it because kubernetes Endpoint is not constantly being updated anymore. It looks weird Pilot building different service accounts on each iteration for `kubernetes.default.svc.cluster.default`.

If there is anything I could change in the code to get more info let me know, it's relatively simple to reproduce it with a custom built pilot.

I also saw that were many changes around locking `*DiscoveryServer` and I want to try `1.2.0-rc.0` and see if I get the same behavior.

**Comment by howardjohn** [comment]
We have an open issue regarding ServiceAccounts for kube-system services, since like you mentioned they sometimes/always(?) don't have a service account associated with them. What is weird is that it is flipping between not having one and having one. Are there two instances of it, and it picks a different one sometimes and not others possibly?

Issue for reference: https://github.com/istio/istio/issues/12108

**Comment by cainelli** [comment]
We do have two instances of the API server, I double checked and the configuration looks the same, none of them with sa.

**Comment by howardjohn** [comment]
I would suspect it is related to the two instances if it is flipping back and forth. Any idea where `calico-node` came from if it isn't in the SA for either of them?

**Comment by cainelli** [comment]
I've been through all the pods checking for the service accounts but the only ones assigned to `calico-nodes` are the calico Pods.

```
$ kubectl  get po -o json --all-namespaces | jq '.items[] | .metadata.namespace + "/" + .metadata.name + ":sa:" + .spec.serviceAccount'  | grep calico-node
"kube-system/calico-node-2gjc7:sa:calico-node"
"kube-system/calico-node-2hjmb:sa:calico-node"
"kube-system/calico-node-4bvc7:sa:calico-node"
"kube-system/calico-node-4gwsk:sa:calico-node"
"kube-system/calico-node-6wf9k:sa:calico-node"
"kube-system/calico-node-6z2h9:sa:calico-node"
"kube-system/calico-node-6zpr4:sa:calico-node"
"kube-system/calico-node-84mj4:sa:calico-node"
"kube-system/calico-node-87ng7:sa:calico-node"
"kube-system/calico-node-9t54m:sa:calico-node"
"kube-system/calico-node-bckph:sa:calico-node"
"kube-system/calico-node-bxqsh:sa:calico-node"
"kube-system/calico-node-cd7wz:sa:calico-node"
"kube-system/calico-node-cx5jv:sa:calico-node"
"kube-system/calico-node-dfpqx:sa:calico-node"
"kube-system/calico-node-f2bcb:sa:calico-node"
"kube-system/calico-node-f2p4m:sa:calico-node"
"kube-system/calico-node-f5fpm:sa:calico-node"
"kube-system/calico-node-f6lhg:sa:calico-node"
"kube-system/calico-node-fln27:sa:calico-node"
"kube-system/calico-node-g5l4p:sa:calico-node"
"kube-system/calico-node-gdrvk:sa:calico-node"
"kube-system/calico-node-gvt2t:sa:calico-node"
"kube-system/calico-node-gxqgw:sa:calico-node"
"kube-system/calico-node-h2qcr:sa:calico-node"
"kube-system/calico-node-hflzw:sa:calico-node"
"kube-system/calico-node-j4t72:sa:calico-node"
"kube-system/calico-node-jnb8b:sa:calico-node"
"kube-system/calico-node-jzwx7:sa:calico-node"
"kube-system/calico-node-kq2xp:sa:calico-node"
"kube-system/calico-node-krbn8:sa:calico-node"
"kube-system/calico-node-l7prd:sa:calico-node"
"kube-system/calico-node-lvmrx:sa:calico-node"
"kube-system/calico-node-mcbsc:sa:calico-node"
"kube-system/calico-node-mfm5f:sa:calico-node"
"kube-system/calico-node-mn6fd:sa:calico-node"
"kube-system/calico-node-nxcw8:sa:calico-node"
"kube-system/calico-node-prlvq:sa:calico-node"
"kube-system/calico-node-qr9br:sa:calico-node"
"kube-system/calico-node-sl5x6:sa:calico-node"
"kube-system/calico-node-slxrs:sa:calico-node"
"kube-system/calico-node-sz742:sa:calico-node"
"kube-system/calico-node-tjtlk:sa:calico-node"
"kube-system/calico-node-v7stx:sa:calico-node"
"kube-system/calico-node-vwr47:sa:calico-node"
"kube-system/calico-node-wm4px:sa:calico-node"
"kube-system/calico-node-xnb22:sa:calico-node"
"kube-system/calico-node-zc4jv:sa:calico-node"
"kube-system/calico-node-zt9qw:sa:calico-node"
```

**Comment by howardjohn** [comment]
@myidpt any ideas how it could be generating a spiffe for the wrong SA?

**Comment by cainelli** [comment]
sharing a bit more on what I've learned since yesterday:

1. It's not exclusively to `calico-node` SA, the flapping service account in `kubernetes.default.svc.cluster.local` also happened with kube-proxy's service account:
```
2019-06-07T09:53:01.682237Z	info	ads	Endpoint updating service account spiffe://cluster.local/ns/kube-system/sa/kube-proxy kubernetes.default.svc.cluster.local
---
2019-06-07T09:53:05.686249Z	info	ads	Endpoint updating service account spiffe://cluster.local/ns/kube-system/sa/ kubernetes.default.svc.cluster.local
```

2. I've upgraded to `1.2.0-rc.0` on the same sandbox cluster and I don't see the flapping SA behavior anymore. In both cases I have the buggy kubernetes service endpoint:

```bash
$ kubectl get ep kubernetes -w
default   kubernetes   10.20.9.86:443   6s
default   kubernetes   10.20.26.106:443   10s
default   kubernetes   10.20.9.86:443   16s
default   kubernetes   10.20.26.106:443   20s
default   kubernetes   10.20.9.86:443   26s
default   kubernetes   10.20.26.106:443   30s
default   kubernetes   10.20.9.86:443   36s
default   kubernetes   10.20.26.106:443   41s
default   kubernetes   10.20.9.86:443   46s
default   kubernetes   10.20.26.106:443   51s
default   kubernetes   10.20.9.86:443   56s
```

> Note: I've deleted the Endpoint so you can see the update frequency by the age column.

**Comment by howardjohn** [comment]
>  I've changed that and by adding --endpoint-reconciler-type=lease to the kubernetes api made things more stable and we see a drastic improve in CPU utilization because it's not doing full push every a couple of sec anymore.

Does this fix not work for the kubernetes service anymore?

**Comment by cainelli** [comment]
It did fixed. Here I'm running without it to be able to reproduce the flapping service account.

**Comment by cainelli** [comment]
To be clear (as I noticed the 1.2 milestone), the flapping service account seems to be fixed in `1.2.0-rc.0` (even with the buggy kubernetes service). However, would be good to understand why it's happening in 1.1.7.

**Comment by hzxuzhonghu** [comment]
Really weird, about this log ,seems the service and the service account do not match.
`				adsLog.Infof("Endpoint updating service account %s %s", e.ServiceAccount, serviceName)
`

>To be clear (as I noticed the 1.2 milestone), the flapping service account seems to be fixed in 1.2.0-rc.0 (even with the buggy kubernetes service). However, would be good to understand why it's happening in 1.1.7.

I think this is fixed by #13689

**Comment by cainelli** [comment]
I saw your PR during my research and even tried to patch my 1.1.7 with it, then I gave up since it looks there are too many other changes to make it work. I wonder if it would make sense have it in 1.1 or accept the fix is only in 1.2.

**Comment by howardjohn** [comment]
@cainelli can you confirm this is now fixed in 1.2? As you mentioned this is pretty challenging to get backported to 1.1 as there have been a lot of changes in the area and [1.1 will be out of life soon](https://discuss.istio.io/t/support-for-istio-1-1-ends-on-september-19th-2019/3489)

**Comment by cainelli** [comment]
Yes it is fixed. I'm closing the issue then and thank you for the help.
